apns.py

```diff
@@ -1,323 +1,420 @@
-            or _get_field(payload[1], 1) != 0x00.to_bytes(1, "big")
-        ):
+        await self._send(payload)
+
+        payload = await self._receive(8)
+
+        if payload.fields_with_id(1)[0].value != b"\x00":
             raise Exception("Failed to connect")
 
-        new_token = _get_field(payload[1], 3)
-        if new_token is not None:
-            self.token = new_token
-        elif token is not None:
-            self.token = token
+        if len(payload.fields_with_id(3)) > 0:
+            new_token = payload.fields_with_id(3)[0].value
         else:
-            raise Exception("No token")
-        
-        logger.debug(f"Recieved connect response with token {b64encode(self.token).decode()}")
+            if token is None:
+                raise Exception("No token received")
+            new_token = token
 
-        return self.token
+        logger.debug(
+            f"Received connect response with token {b64encode(new_token).decode()}"
+        )
 
-    def filter(self, topics: list[str]):
+        return new_token
+
+    old_topics = []
+    async def filter(self, topics: list[str]):
+        """Sends the APNs filter message"""
+        if topics == self.old_topics:
+            return
+        topics = list(set(topics + self.old_topics))
+        self.old_topics = topics
         logger.debug(f"Sending filter message with topics {topics}")
-        fields = [(1, self.token)]
 
-        for topic in topics:
-            fields.append((2, sha1(topic.encode()).digest()))
+        payload = APNSPayload(9, [APNSField(1, self.credentials.token)])
 
-        payload = _serialize_payload(9, fields)
+        for topic in topics:
+            payload.fields.append(APNSField(2, sha1(topic.encode()).digest()))
 
-        self.sock.write(payload)
+        await payload.write_to_stream(self.sock)
 
-    def send_message(self, topic: str, payload: str, id=None):
-        logger.debug(f"Sending message to topic {topic} with payload {payload}")
+    async def send_notification(self, topic: str, payload: bytes, id=None):
+        """Sends a notification to the APNs server"""
+        logger.debug(f"Sending notification to topic {topic}")
         if id is None:
             id = random.randbytes(4)
 
-        payload = _serialize_payload(
-            0x0A,
+        p = APNSPayload(
+            0xA,
             [
-                (4, id),
-                (1, sha1(topic.encode()).digest()),
-                (2, self.token),
-                (3, payload),
+                APNSField(4, id),
+                APNSField(1, sha1(topic.encode()).digest()),
+                APNSField(2, self.credentials.token),
+                APNSField(3, payload),
             ],
         )
 
-        self.sock.write(payload)
+        await self._send(p)
 
         # Wait for ACK
-        payload = self.incoming_queue.wait_pop_find(lambda i: i[0] == 0x0B)
+        r = await self._receive(0xB)
+
+        # TODO: Check ACK code
```

Comment:
is this something that needs to be taken care of? and if it doesn't need to be dealt with right now, could you make it a bit more specific about what is needed?

Reply:
there is a status code in the ACK response, it should technically be checked to be 00 and throw an exception if it is not